app/_src/kubernetes-ingress-controller/guides/security/verify-upstream-tls.md

```diff
@@ -0,0 +1,392 @@
+
+Verify connectivity by issuing an HTTP request to proxy. The service serves HTTPS but {{ site.base_gateway }} initiates
+the connection and proxies it as HTTP in this case, thus the request should be made over HTTP. The `Host` header is
+required to match the hostname of the service.
+
+```shell
+curl -H Host:kong.example $PROXY_IP/echo
+```
+
+You should see a response similar to this:
+
+```text
+Welcome, you are connected to node orbstack.
+Running on Pod echo-bd94b7dcc-qxs2b.
+In namespace default.
+With IP address 192.168.194.9.
+Through HTTPS connection.
+```
+
+That means the service is up and running and {{ site.base_gateway }} connects to it successfully over HTTPS, _without
+verification_.
+
+## Configure {{ site.base_gateway }} to verify the upstream TLS certificate
+
+### Enable TLS verification
+
+To configure {{ site.base_gateway }} to verify the certificate of the upstream service, we need to annotate the
+service accordingly.
+
+```shell
+kubectl annotate service echo konghq.com/tls-verify=true
+```
+
+The results should look like this.
+
+```text
+service/echo annotated
+```
+
+Now, when you issue the same request as before, you should see an error similar to this.
+
+```shell
+curl -H Host:kong.example $PROXY_IP/echo
+```
+
+```text
+{
+  "message":"An invalid response was received from the upstream server",
+  "request_id":"e2b3182856c96c23d61e880d0a28012f"
+}
+```
+
+If you inspect {{ site.base_gateway }}'s container logs, you should see an error indicating an issue with TLS handshake.
+
+```shell
+ kubectl logs -n kong deploy/kong-gateway | grep "GET /echo"
+```
+
+```text
+2024/11/28 18:04:05 [error] 1280#0: *20004 upstream SSL certificate verify error: (20:unable to get local issuer certificate) while SSL handshaking to upstream, client: 192.168.194.1, server: kong, request: "GET /echo HTTP/1.1", upstream: "https://192.168.194.9:443/", host: "kong.example", request_id: "e2b3182856c96c23d61e880d0a28012f"
+192.168.194.1 - - [28/Nov/2024:18:04:05 +0000] "GET /echo HTTP/1.1" 502 126 "-" "curl/8.7.1" kong_request_id: "e2b3182856c96c23d61e880d0a28012f"
+```
+
+{{ site.base_gateway }} is now verifying the certificate of the upstream service and rejecting the connection because
+the certificate is not trusted.
+
+### Add the root CA certificate to {{ site.base_gateway }}
+
+That can be fixed by adding the root CA certificate to the {{ site.base_gateway }}'s CA certificates and associating
+it with the service.
+
+First, create a secret with the root CA certificate.
+
+{% include /md/kic/ca-certificates-note.md %}
+
+```shell
+echo '
+apiVersion: v1
+kind: Secret
+metadata:
+  name: root-ca
+  labels:
+    konghq.com/ca-cert: "true" # This label is required for the CA certificate to be recognized by Kong
+  annotations:
+    kubernetes.io/ingress.class: kong
+data:
+  cert: '$(base64 -w0 ./certs/root.crt)'
+  # An arbitrary ID for the certificate 
+  id: '$(printf "bf6e0f14-78cd-45ad-9325-87ec7ef7b890" | base64 -w0)'
```

Comment:
Odd:
```
$ base64 -w0
base64: invalid option -- w
Usage:	base64 [-Ddh] [-b num] [-i in_file] [-o out_file]
  -b, --break    break encoded string into num character lines
  -Dd, --decode   decodes input
  -h, --help     display this message
  -i, --input    input file (default: "-" for stdin)
  -o, --output   output file (default: "-" for stdout)
$ uname -o
Darwin
$ which base64
/usr/bin/base64
$ uname -v
Darwin Kernel Version 23.6.0: Mon Jul 29 21:14:30 PDT 2024; root:xnu-10063.141.2~1/RELEASE_ARM64_T6000
```

Reply:
Alright, it's because I'm using `base64` from https://formulae.brew.sh/formula/coreutils.

```
$ base64 --help                                                                                                                                                                            <<<
Usage: base64 [OPTION]... [FILE]
Base64 encode or decode FILE, or standard input, to standard output.

With no FILE, or when FILE is -, read standard input.

Mandatory arguments to long options are mandatory for short options too.
  -d, --decode          decode data
  -i, --ignore-garbage  when decoding, ignore non-alphabet characters
  -w, --wrap=COLS       wrap encoded lines after COLS character (default 76).
                          Use 0 to disable line wrapping
      --help        display this help and exit
      --version     output version information and exit

The data are encoded as described for the base64 alphabet in RFC 4648.
When decoding, the input may contain newlines in addition to the bytes of
the formal base64 alphabet.  Use --ignore-garbage to attempt to recover
from any other non-alphabet bytes in the encoded stream.

GNU coreutils online help: <https://www.gnu.org/software/coreutils/>
Report any translation bugs to <https://translationproject.org/team/>
Full documentation <https://www.gnu.org/software/coreutils/base64>
or available locally via: info '(coreutils) base64 invocation'
$ which base64                                                                                                                                                                             <<<
/opt/homebrew/bin/gbase64
$ brew ls coreutils | grep base64                                                                                                                                                          <<<
/opt/homebrew/Cellar/coreutils/9.5/bin/gbase64
/opt/homebrew/Cellar/coreutils/9.5/libexec/gnubin/base64
/opt/homebrew/Cellar/coreutils/9.5/libexec/gnuman/man1/base64.1
/opt/homebrew/Cellar/coreutils/9.5/share/man/man1/gbase64.1
```

```
/usr/bin/base64 -w0 CHANGELOG.md
base64: invalid argument CHANGELOG.md
Usage:  base64 [-Ddh] [-b num] [-i in_file] [-o out_file]
  -b, --break       break encoded output up into lines of length num
  -D, -d, --decode  decode input
  -h, --help        display this message
  -i, --input       input file (default: "-" for stdin)
  -o, --output      output file (default: "-" for stdout)
```

I'll update the guide for this to be platform agnostic.
```

Reply:
So the changes are already in place under: https://deploy-preview-8201--kongdocs.netlify.app/kubernetes-ingress-controller/unreleased/guides/security/verify-upstream-tls/#add-the-root-ca-certificate-to-kong-gateway

This section now uses `kubectl create (secret|configmap)`.

Reply:
The steps to create `Secret`/`ConfigMap` do not use `base64` now. PTAL